File: Android/crud-ng/src/main/java/sevenguis/crudng/MainActivity.java
package sevenguis.crudng;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Dialog;
import android.app.DialogFragment;
import android.app.FragmentManager;
import android.content.DialogInterface;
import android.graphics.Color;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import butterknife.ButterKnife;
import butterknife.InjectView;
import butterknife.OnClick;
import butterknife.OnItemClick;
import butterknife.OnTextChanged;

import static butterknife.ButterKnife.findById;

/*
Refer to the crud subproject for various comments.
 */
public class MainActivity extends Activity {

    public static class State extends StateFragment {
        FilterableList<String> filterableList;
        IntHolder selectedIndex;
    }

    private State state;
    private ArrayAdapter<String> adapter;
    private PrefixPredicate predicate;

    @InjectView(R.id.prefix) TextView prefix;
    @InjectView(R.id.entries) ListView entries;
    @InjectView(R.id.create) Button create;
    @InjectView(R.id.update) Button update;
    @InjectView(R.id.delete) Button delete;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.inject(this);

        FragmentManager fm = getFragmentManager();
        state = (State) fm.findFragmentByTag("state");
        if (state == null) {
            state = new State();
            fm.beginTransaction().add(state, "state").commit();
            List<String> externDb;
            externDb = new ArrayList<String>();
            externDb.add("Emil, Hans");
            externDb.add("Mustermann, Max");
            externDb.add("Tisch, Roman");
            state.filterableList = new FilterableList<String>(externDb);
            state.selectedIndex = new IntHolder(-1);
        }

        updateButtonState();

        predicate = new PrefixPredicate(prefix.getText().toString());

        adapter = new ArrayAdapter<String>(getApplicationContext(), android.R.layout.simple_list_item_1, state.filterableList.getFiltered()) {
            @Override
            public View getView(int position, View convertView, ViewGroup parent) {
                TextView itemView = (TextView) super.getView(position, convertView, parent);
                itemView.setTextColor(Color.BLACK); // Interestingly, this is not needed without ButterKnife...
                if (state.selectedIndex.value == state.filterableList.orig(position))
                    itemView.setBackgroundColor(Color.CYAN);
                else
                    itemView.setBackgroundColor(Color.TRANSPARENT);
                return itemView;
            }
        };
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        entries.setAdapter(adapter);
    }

    @OnTextChanged(R.id.prefix) void filter(CharSequence s) {
        predicate.prefix = s.toString();
        state.filterableList.filter(predicate);
        adapter.notifyDataSetChanged();
        updateButtonState();
    }

    @OnItemClick(R.id.entries) void select(int pos) {
        if (state.selectedIndex.value == state.filterableList.orig(pos)) {
            state.selectedIndex.value = -1;
        } else {
            state.selectedIndex.value = state.filterableList.orig(pos);
        }
        adapter.notifyDataSetChanged();
        updateButtonState();
    }

    @OnClick(R.id.create) void create() { showNameDialog(false); }
    @OnClick(R.id.update) void update() { showNameDialog(true); }
    @OnClick(R.id.delete) void delete() {
        state.filterableList.delete(state.selectedIndex.value);
        adapter.notifyDataSetChanged();
        state.selectedIndex.value = -1;
        updateButtonState();
    }

    private void showNameDialog(final boolean update) {
        Bundle args = new Bundle();
        args.putBoolean("update", update);
        NameDialog nd = new NameDialog();
        nd.setArguments(args);
        nd.show(getFragmentManager(), update ? "UpdateDialog" : "CreateDialog");
    }

    private void updateButtonState() {
        if (state.filterableList.filt(state.selectedIndex.value) == -1) {
            update.setEnabled(false);
            delete.setEnabled(false);
        } else {
            update.setEnabled(true);
            delete.setEnabled(true);
        }
    }

    public static class NameDialog extends DialogFragment {
        @Override
        public Dialog onCreateDialog(Bundle savedInstanceState) {
            final boolean update = getArguments().getBoolean("update");
            final MainActivity activity = (MainActivity) getActivity();
            final FilterableList<String> filterableList = activity.state.filterableList;
            final IntHolder selectedIndex = activity.state.selectedIndex;
            final ArrayAdapter<String> adapter = activity.adapter;
            final LayoutInflater inflater = activity.getLayoutInflater();
            final View nameDialog = inflater.inflate(R.layout.name_dialog, null);
            final EditText name = findById(nameDialog, R.id.name);
            final EditText surname = findById(nameDialog, R.id.surname);
            if (update) {
                String[] fullname = filterableList.get(selectedIndex.value).split(",");
                name.setText(fullname[1].trim());
                surname.setText(fullname[0].trim());
            }

            AlertDialog.Builder builder = new AlertDialog.Builder(activity);
            builder.setTitle("Create Entry")
                    .setView(nameDialog)
                    .setPositiveButton("Ok", new DialogInterface.OnClickListener() {
                        public void onClick(DialogInterface dialog, int id) {
                            String fullname = surname.getText().toString() + ", " +
                                    name.getText().toString();
                            if (update) {
                                filterableList.update(fullname, selectedIndex.value);
                            } else {
                                filterableList.create(fullname);
                            }
                            adapter.notifyDataSetChanged();
                        }
                    })
                    .setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
                        public void onClick(DialogInterface dialog, int id) {
                            // Do nothing
                        }
                    });
            return builder.create();
        }
    }

    private static class IntHolder {
        public int value;
        public IntHolder(int v) { value = v; }
    }

    private static class PrefixPredicate implements FilterableList.Predicate<String> {
        String prefix;
        PrefixPredicate(String prefix) { this.prefix = prefix; }
        @Override
        public boolean f(String element) {
            return element.toLowerCase().startsWith(prefix);
        }
    }

}

File: Android/crud-kt/src/main/kotlin/MainActivity.kt
package sevenguis.crudkt

import android.app.Activity
import android.os.Bundle
import sevenguis.crudkt.R
import kotlin.properties.Delegates
import android.widget.ArrayAdapter
import android.widget.TextView
import android.widget.ListView
import android.widget.Button
import android.view.View
import android.view.ViewGroup
import android.graphics.Color
import android.app.DialogFragment
import android.app.Dialog
import android.widget.EditText
import android.app.AlertDialog

/*
Here I actually don't like Kotlin's strictness when it comes to enforcing null checks.
As in Android things aren't typically initialized with a constructor but with certain
lifecycle methods the compiler cannot easily be shown that something is certainly not
null. But then the !!s are inconvient and so you end up with Delegates.notNull() anyway
and that's just verbose. Why not reuse the platform types syntax for Kotlin types?
Sometimes you just want to avoid the over-strictness of a static typing regime because
you know more than the compiler and you value convenience for certain situations
higher than convincing a machine.
(In general, I'm more and more of the opinion that static null checking should be
opt-in instead of opt-out... For example MyType is like in Java no checks, MyType!
means that compiler should check that it is never null and MyType? means that compiler
analyzes nullability)

So using Kotlin without annotations one could (almost) achieve the convenience of
Butterknife for Java. Kotlin's general good features (e.g. extension functions)
help Android development of course. Callback-heavy Android code could greatly benefit
from Kotlin (e.g. RxJava). Nullability analysis is inconvenient for Android.
There are current problems with Kotlin and libraries like Dagger.
I have the strong feeling that metalinguistic (syntactic) abstraction could
considerably reduce Android boilerplate e.g. with respect to generically patching
many inconveniences when working with fragments (say providing arguments).
 */
public class MainActivity : Activity() {

    class State : StateFragment() {
        // Can't we just do FilterableList<String>! as a shortcut?
        var filterableList: FilterableList<String> by Delegates.notNull()
        var selectedIndex = -1
    }

    var state: State by Delegates.notNull()
    var adapter: ArrayAdapter<String> by Delegates.notNull()

    val prefix: TextView by bindView(R.id.prefix)
    val entries: ListView by bindView(R.id.entries)
    val create: Button by bindView(R.id.create)
    val update: Button by bindView(R.id.update)
    val delete: Button by bindView(R.id.delete)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val fm = getFragmentManager()
        val s = fm.findFragmentByTag("state") as State?
        if (s == null) {
            state = State()
            val externDb = arrayListOf("Emil, Hans", "Mustermann, Max", "Tisch, Roman")
            state.filterableList = FilterableList(externDb)
            state.selectedIndex = -1
            fm.beginTransaction().add(state, "state").commit()
        } else {
            state = s
        }

        updateButtonState()

        val predicate = PrefixPredicate(prefix.getText().toString())

        adapter = object : ArrayAdapter<String>(getApplicationContext(), android.R.layout.simple_list_item_1, state.filterableList.filtered) {
            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View? {
                val itemView = super.getView(position, convertView, parent) as TextView
                itemView.setTextColor(Color.BLACK)
                if (state.selectedIndex == state.filterableList.orig(position))
                    itemView.setBackgroundColor(Color.CYAN)
                else
                    itemView.setBackgroundColor(Color.TRANSPARENT)
                return itemView
            }
        }
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        entries.setAdapter(adapter)

        prefix onTextChanged { s ->
            predicate.prefix = s.toString()
            // Why no SAM conversion in the other direction here?
            //state.filterableList.filter(predicate)
            // Or would this make sense
            //state.filterableList.filter(::predicate)
            // This begs for eta reduction :)
            state.filterableList.filter { s -> predicate(s) }
            adapter.notifyDataSetChanged()
            updateButtonState()
        }

        entries onItemClick { pos ->
            if (state.selectedIndex == state.filterableList.orig(pos)) {
                state.selectedIndex = -1
            } else {
                state.selectedIndex = state.filterableList.orig(pos)
            }
            adapter.notifyDataSetChanged()
            updateButtonState()
        }

        create onClick { showNameDialog(false) }
        update onClick { showNameDialog(true) }
        delete onClick {
            state.filterableList.delete(state.selectedIndex)
            adapter.notifyDataSetChanged()
            state.selectedIndex = -1
            updateButtonState()
        }
    }

    fun showNameDialog(update: Boolean) {
        val args = Bundle()
        args.putBoolean("update", update)
        val nd = NameDialog()
        nd.setArguments(args)
        nd.show(getFragmentManager(), if (update) "UpdateDialog" else "CreateDialog")
    }

    fun updateButtonState() {
        if (state.filterableList.filt(state.selectedIndex) == -1) {
            update.setEnabled(false)
            delete.setEnabled(false)
        } else {
            update.setEnabled(true)
            delete.setEnabled(true)
        }
    }

    class NameDialog : DialogFragment() {
        override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
            val update = getArguments().getBoolean("update")
            val activity = getActivity() as MainActivity
            val filterableList = activity.state.filterableList
            val selectedIndex = activity.state.selectedIndex
            val adapter = activity.adapter
            val inflater = activity.getLayoutInflater()
            val view = inflater.inflate(R.layout.name_dialog, null)
            val name = view.findViewById(R.id.name) as EditText
            val surname = view.findViewById(R.id.surname) as EditText
            if (update) {
                val fullname = filterableList[selectedIndex].split(",")
                name.setText(fullname[1].trim())
                surname.setText(fullname[0].trim())
            }

            val builder = AlertDialog.Builder(activity)
            builder.setTitle("Create Entry")
                .setView(view)
                .setPositiveButton("Ok") { dialog, id ->
                    val fullname = "${surname.getText()}, ${name.getText()}"
                    if (update)
                        filterableList.update(fullname, selectedIndex)
                    else
                        filterableList.create(fullname)
                    adapter.notifyDataSetChanged()
                }
            .setNegativeButton("Cancel") { dialog, id -> /* Nothing */ }
            return builder.create()
        }
    }

    class PrefixPredicate(prefix: String) {
        var prefix: String = prefix
            get() = $prefix
            set(s: String) { $prefix = s.toLowerCase() }
        fun invoke(s: String): Boolean = s.toLowerCase().startsWith(prefix)
    }

}